src/nvim/eval/funcs.c

```diff
@@ -1692,7 +1692,9 @@ static void f_exists(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
   int n = false;
 
   const char *p = tv_get_string(&argvars[0]);
-  if (*p == '$') {  // Environment variable.
+  if (strncmp(p, "v:lua.", 6) == 0 && nlua_func_exists(p + 6)) {
```

Comment:
this will be very uncommon so it should be the last condition, not the first. (micro-optimization, but still...)